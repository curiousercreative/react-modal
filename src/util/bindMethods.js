// bind all methods except whose that match this regex (react lifecycle methods)
const REGEX = /^(component.*|render)$/;

/**
 * bind the `this` keyword in event handling class methods.
 * This is a side effects function, mutates instance parameter
 * @function bindMethodsTo
 * @param  {object} instance - object whose methods to look over and bind to itself
 */
export default function bindMethods (instance) {
  // look over all properties and methods of this object
  getPropertyNames(instance)
    // filter to only methods (eliminate properties)
    .filter(methodName => typeof instance[methodName] === 'function')
    // filter to only look at methods that don't match of regex
    .filter(methodName => !methodName.match(REGEX))
    // bind these methods to the class instance
    .forEach(methodName => {
      instance[methodName] = instance[methodName].bind(instance);
    });
}

function getPropertyNames (instance) {
  let propertyNames = [];

  // starting with this instance and crawling our way up the prototype chain...
  // add all property names to an array until we hit the React.Component class
  for (let proto = instance; !proto.hasOwnProperty('isReactComponent'); proto = Object.getPrototypeOf(proto)) {
    propertyNames = propertyNames.concat(Object.getOwnPropertyNames(proto));
  }

  return propertyNames;
}
