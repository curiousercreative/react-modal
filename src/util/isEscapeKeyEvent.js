/**
 * isEscapeKeyEvent
 * @param  {Event}  e
 * @return {Boolean}
 */
export default function isEscapeKeyEvent (e) {
  const key = e.key || e.keyCode;

  return [ 'Escape', 'Esc', 27 ].includes(key);
}
