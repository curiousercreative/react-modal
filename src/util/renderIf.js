const defaultNeg = () => null;

/**
 * @param  {boolean} expression
 * @param  {function} affirmative
 * @param  {function} [negative]
 * @return {jsx}
 */
export default function renderIf (expression, affirmative, negative = defaultNeg) {
  return expression ? affirmative() : negative();
}
