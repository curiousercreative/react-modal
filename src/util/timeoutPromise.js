/**
 * create a promise that resolves after provided amount of time has lapsed
 * @function timeoutPromise
 * @param  {number} ms - amount of time to wait (in milliseconds)
 * @return {promise}
 */
export default function timeoutPromise (ms=0) {
  return new Promise(resolve => setTimeout(resolve, ms));
}
