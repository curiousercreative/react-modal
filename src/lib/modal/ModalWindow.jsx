import React from 'react';

import bindMethods from '../../util/bindMethods';
import isEscapeKeyEvent from '../../util/isEscapeKeyEvent';
import renderIf from '../../util/renderIf';
import timeoutPromise from '../../util/timeoutPromise';
import * as modal from './';
import { addEventListener, body, documentElement, hasScrollbar, querySelector, removeEventListener } from '../dom.js';

/**
 * @property  {jsx} children
 * @property  {string} [className]
 * @property  {boolean} [centered = true]
 * @property  {boolean} [dismissable = true] - whether modal can be dismissed
 * @return  {jsx} div
 */
export default class ModalWindow extends React.Component {
  static defaultProps = {
    className: '',
    centered: true,
    dismissable: true,
  };

  state = {
    open: false,
  };
  windowRef = React.createRef();

  constructor (...args) {
    super(...args);
    bindMethods(this);
  }

  componentDidMount () {
    // start the window animation
    this.setState({ open: true });

    // darken background
    // async to allow .modal-container:empty style to clear before applying .modal-container--open
    setTimeout(() => {
      // scroll lock but preserve vertical scrollbar on browser window
      if (hasScrollbar()) {
        body.style.top = `-${documentElement.scrollTop}px`;
        body.classList.add('modal-open-with-scrollbar');
      }
      body.classList.add('modal-open');
      querySelector('.modal-container').classList.add('modal-container--open');

      // listen for clicks outside of us
      addEventListener('click', this.onClicks);
      addEventListener('keyup', this.onKeys);
    }, 0);
  }

  componentWillUnmount () {
    // restore and cleanup after scroll locking
    querySelector('.modal-container').classList.remove('modal-container--open');
    body.classList.remove('modal-open-with-scrollbar', 'modal-open');
    documentElement.scrollTop = Math.abs(parseInt(body.style.top, 10));
    body.style.top = null;

    // stop listening
    removeEventListener('click', this.onClicks);
    removeEventListener('keyup', this.onKeys);
  }

  handleClickClose (e) {
    e.preventDefault();
    modal.close();
  }

  onClicks (e) {
    // close if clicking outside
    if (
      !e.defaultPrevented
      && this.props.dismissable
      && !this.windowRef.current.contains(e.target)
    ) modal.close();
  }

  onKeys (e) {
    if (
      !e.defaultPrevented
      && this.props.dismissable
      && isEscapeKeyEvent(e)
    ) modal.close();
  }

  prepareToClose () {
    // lighten background
    querySelector('.modal-container').classList.remove('modal-container--open');

    // start the closing animation
    this.setState({ open: false });

    return timeoutPromise(modal.ANIMATION_DURATION);
  }

  render () {
    let classes = this.props.className.split(' ').concat([ 'modal', 'modal__window' ]);

    if (this.state.open) classes.push('modal--open');
    if (this.props.centered) classes.push('modal__window--centered');

    return (
      <div aria-modal="true" className={classes.join(' ')} ref={this.windowRef} role="dialog">
        {renderIf(this.props.dismissable, () => (
          <button className="modal__close" onClick={this.handleClickClose} type="button" />
        ))}
        {this.props.children}
      </div>
    );
  }
}
