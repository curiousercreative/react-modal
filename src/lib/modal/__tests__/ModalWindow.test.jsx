import React from 'react';
import renderer from 'react-test-renderer';
import { fireEvent, waitForElementToBeRemoved } from '@testing-library/react';

import * as modal from '../';
import ModalWindow from '../ModalWindow';
import { setup, teardown } from './util';

beforeAll(setup);
afterEach(teardown);

describe('ModalWindow tests', () => {
  test('ModalWindow should match snapshot', () => {
    const instance = renderer.create(<ModalWindow>Hey there!</ModalWindow>);
    expect(instance.toJSON()).toMatchSnapshot();
    // bizaare that we have to clean-up...
    instance.unmount();
  });

  test('ModalWindow props should be passed via modal.open', () => {
    return modal
      .open(<div>Hello</div>, { bogus: 'yes' })
      .then(instance => expect(instance.props.bogus).toEqual('yes'));
  });

  test('ModalWindow prop.dismissable should default to true', () => {
    return modal
      .open(<div>Hello</div>)
      .then(instance => expect(instance.props.dismissable).toBeTruthy());
  });

  test('ModalWindow should render close button by default', () => {
    return modal
      .open(<div>Hello</div>)
      .then(() => expect(document.querySelector('.modal__close')).toBeTruthy());
  });

  test('ModalWindow should listen for close button clicks by default', () => {
    return modal
      .open(<div>Hello</div>)
      .then(() => fireEvent.click(document.querySelector('.modal__close')))
      .then(() => waitForElementToBeRemoved(() => document.querySelector('.modal')))
      .then(() => expect(document.querySelector('.modal')).toBeFalsy());
  });

  test('ModalWindow should listen for clicks outside of modal by default', () => {
    return modal
      .open(<div>Hello</div>)
      .then(() => fireEvent.click(document.querySelector('.modal-container')))
      .then(() => waitForElementToBeRemoved(() => document.querySelector('.modal')))
      .then(() => expect(document.querySelector('.modal')).toBeFalsy());
  });

  test('ModalWindow should listen for escape key by default', () => {
    return modal
      .open(<div>Hello</div>)
      // https://testing-library.com/docs/dom-testing-library/api-events#fireeventeventname
      .then(() => fireEvent.keyUp(document.body, { code: 'Escape', key: 'Escape', keyCode: 27 }))
      .then(() => waitForElementToBeRemoved(() => document.querySelector('.modal')))
      .then(() => expect(document.querySelector('.modal')).toBeFalsy());
  });

  test('ModalWindow prop.dismissable=false should disable close button rendering', () => {
    return modal
      .open(<div>Hello</div>, { dismissable: false })
      .then(() => expect(document.querySelector('.modal__close')).toBeFalsy());
  });

  test('ModalWindow prop.dismissable=false should disable clicks outside of modal closing', () => {
    return modal
      .open(<div>Hello</div>, { dismissable: false })
      .then(() => fireEvent.click(document.querySelector('.modal-container')))
      .then(() => expect(document.querySelector('.modal--open')).toBeTruthy());
  });

  test('ModalWindow prop.dismissable=false should disable escape key listener', () => {
    return modal
      .open(<div>Hello</div>, { dismissable: false })
      .then(() => expect(document.querySelector('.modal--open')).toBeTruthy())
      // https://testing-library.com/docs/dom-testing-library/api-events#fireeventeventname
      .then(() => fireEvent.keyUp(document.body, { code: 'Escape', key: 'Escape', keyCode: 27 }))
      .then(() => expect(document.querySelector('.modal--open')).toBeTruthy());
  });
});
