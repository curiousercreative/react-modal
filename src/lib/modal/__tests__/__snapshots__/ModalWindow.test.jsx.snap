// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalWindow tests ModalWindow should match snapshot 1`] = `
<div
  aria-modal="true"
  className=" modal modal__window modal--open modal__window--centered"
  role="dialog"
>
  <button
    className="modal__close"
    onClick={[Function]}
    type="button"
  />
  Hey there!
</div>
`;
